### ISSUE TITLE
Handle chain reorgs in newHeads event

### ISSUE THREAD

**Issue by benjamincburns** [trufflesuite/ganache-core#91]
See expected behavior here. Today Ganache doesn't do this.
https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB#newheads

**Comment by snario** [comment]
Any thought been put towards this?

**Comment by nicholasjpaterno** [comment]
@snario yes definitely! However, it seems this has been a lower priority item for a while now.  With that said, I see this as more of a `bug` than an `enhancement` so I'm going to relabel and bump the priority on this now!

In the meantime, is it possible you can share a little more about your use case?

**Comment by snario** [comment]
Since state channels rely heavily on the on-chain state and events that were emitted leading up to it, we need to have a very detailed snapshot of what the latest state of the chain is as well as updates with regards to what blocks (and the events contained within them) make up that state. 

It seems to me like `newHeads` is used to help with this. We're not sure necessarily if we're going to use `newHeads` or not though. In any case though, we're trying to develop a robust way of matching up state and events into a detailed and reliable object that you can be confident about being the "latest state" of the chain.

**Comment by Troublor** [comment]
Hey, are you guys working on solving this issue?

**Comment by davidmurdoch** [comment]
We're in the middle of a large refactor of ganache's internals. I'm hopefully this new architecture will make simulating chain reorgs easier.

**Comment by 717a56e1** [comment]
Has this been started? Or the large refactor is still in progress?

**Comment by raphael-verdier** [comment]
Hey, did this end up being pushed back for later? I see that the code was somewhat on its way.

**Comment by davidmurdoch** [comment]
It is currently stalled due to concerns raised here: https://github.com/trufflesuite/ganache/pull/3054#pullrequestreview-977468872
